adr/ADR-19.md

```diff
@@ -0,0 +1,273 @@
+# JetStream based Object Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-11-03|
+|Author  |@scottf|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, objectstore|
+
+## Context
+
+This document describes a design of a JetStream backed object store.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+- Represent an object store.
+- Store a large quantity of related bytes in chunks as a single object.
+- Retrieve all the bytes from a single object
+- Store meta data regarding each object
+- Store multiple objects in a single store  
+- Ability to specify chunk size
+- Ability to delete an object
+- Ability to understand the state of the object store.
+
+Possible future features
+
+- Event Notifications (add/delete/lock)
+- Locking
+- Archiving (tiered storage)
+- Searching/Indexing (tagging)
+- Versioning / Revisions
+- Overriding digest algorithm 
+- Capturing Content-Type (mime type)
+- Per chunk Content-Encoding (i.e. gzip)
+
+## Naming Specification
+
+Protocol Naming Conventions are fully defined in [ADR-6](ADR-6.md)
+
+### Object Store
+The object store name or bucket name (`os-bucket-name`) will be used to formulate a stream name and is specified as: `restricted-term` or 1 or more of `A-Z, a-z, 0-9, dash, underscore`
+
+### Objects
+An individual object name is custom url encoding to form a portion of a subject (`os-object-name`).
+There are essentially no restrictions on the object name except that it has to be a valid filename across operating systems, so it is specified as: 
+one or more file `filename-full`, maximum 255 characters where `filename-full` is defined as 
+`printable except asterisk, lt, gt, colon, double-quote, fwd-slash, backslash, pipe, question-mark, ampersand`.
+
+The custom portion of the url encoding is that dot `.` must also be encoded as `%2e`, which is not required in standard url encoding.
+
+### Chunk Ids
+Chunk ids (`chunk-id`) should be a nuid.
+
+### Component Templates
+
+| Component | Template |
+| --- | --- |
+| Stream Name | `OBJ_<os-bucket-name>` |
+| Object Info Stream subject | `$O.<os-bucket-name>.M.>` |
+| Chunk Stream subject | `$O.<os-bucket-name>.C.>` |
+| Object Info message subject | `$O.<os-bucket-name>.M.<os-object-name>` |
```

Comment:
not quite sure why we are storing the object name here - effectively if the stream supported multiple files, this would be a really good idea, because then you could list the objects by scanning this subject pattern.

I am thinking that the name of the object should be the bucket name and whether that is a file or not is orthogonal to the functioning of the object store.

Reply:
1 bucket can hold many files

Reply:
I agree, `<os-bucket-name>` is redundant after the stream name. They are only in that exact bucket - or are they. What if those messages get moved to a mirror which could be another stream name? Is that right?

Reply:
`<os-object-name>` is to allow multiple files per stream (bucket / store)